docs/hugo/content/design/secrets.md

```diff
@@ -0,0 +1,669 @@
+used across all of the providers.
+
+The destination of the `writeConnectionSecretToRef` is currently always a Kubernetes secret, but there is an [open issue](https://github.com/crossplane/crossplane/issues/2366) 
+requesting pluggable secret stores.
+
+**User provided secrets** seem to be 
+[automatically generated](https://github.com/crossplane/provider-azure/blob/f37af6dd4f9d7d10a14caf8270d38078aee06bbe/pkg/controller/database/mysqlserver/managed.go#L157)
+by Crossplane.
+
+### ACK
+
+**User provided secrets** are provided via a [SecretKeyRef](https://github.com/aws-controllers-k8s/runtime/blob/main/apis/core/v1alpha1/secret.go). 
+This allows cross-namespace references to a secret. The specific `Key` of the secret is selected with a `Key string` parameter.
+
+It looks like there may [not currently be support](https://github.com/aws-controllers-k8s/community/issues/700) for key updates/rollover.
+
+**AWS generated secrets** I can't find any examples of. They do not seem to classify "endpoints" as a secret, as shown by
+[opensearchservice endpoint](https://aws-controllers-k8s.github.io/community/reference/opensearchservice/v1alpha1/domain/) and
+[cluster endpoint](https://aws-controllers-k8s.github.io/community/reference/eks/v1alpha1/cluster/).
+
+## Proposal
+
+### User specified secrets
+
+User specified secrets will be detected and transformed from `string` to a `SecretRef`:
+
+```go
+type SecretRef struct {
+    // SecretName is the name of the secret. The secret must be in the same namespace as the resource.
+    SecretName string
+
+    // Key is the key in the secret to use.
+    Key string
+}
+```
+
+Detection will be done with a combination of:
+1. Additions to ASO's configuration file to flag particular fields as secret.
+2. Using the `"format": "password"` data from Swagger, such as that used by 
+   [MySQL Flexible Server](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/mysql/resource-manager/Microsoft.DBforMySQL/stable/2021-05-01/mysql.json#L1662). 
+   Note that not all specs have this, for example
+   [VMSS does not](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json#L9397).
+3. Using the `x-ms-secret` annotation, which I assume has the same meaning as `"format": "password"` although it's not actually exactly documented anywhere.
+
+This is a place where we can push changes upstream to flag things as passwords if they're not being flagged.
+
+#### Lifecycle
+Since these secrets were created _by the user_, the user owns the lifecycle of these secrets. They could be using the same secret across many resources, or 
+intending to use this secret on a resource they have not created yet. As such, the lifecycle of these secrets must be controlled by the user.
+
+Rollover will be supported by triggering events on the associated resource when the secret is modified. Since multiple custom resources might be using 
```

Comment:
It will depend on the actual service/API in question. In the case of most databases that I have seen, their administrator account only has a single password, so any time that password is changed for rollover purposes anything using that account would be impacted yes. There's not a lot we can do to fix that though. The DB itself would need to support multiple passwords so that they could roll one and then the other (similar to how `StorageAccount` or `CosmosDB` support multiple keys today).

With that said, I suspect that's not a _huge_ risk. Most places actually using databases in production won't be using the the administrator account for their applications, they would have dedicated accounts for that. If/when we support individual database accounts (a dataplane API) this will come up again but from my reading on the internet the "best practice" seems to be that if your application fails with an auth error to fail out and start again by reading your secret from the secret store - which would be the Kubernetes secret in this case and so while there would be an interruption when rolling the password it wouldn't be a major one - pods would restart, get injected with the new password, and pick back up where they left off. Technically speaking if the password update takes a long time the secret in Kubernetes can be "too new" (as Azure hasn't accepted the new one yet) and if pods restarted early they'd get the wrong secret and fail for a bit. 

That's a really long-winded way of saying: No, I don't think we need to worry about coordination or timing here. If the user asks us to update the password (by changing it in the Kubernetes secret), we should just do it. We aren't going to be able to guarantee 100% uptime unless the service in question supports multiple keys/passwords/secrets, and if the service _does_ support that, we should be able to do what we've designed here and we will automatically get those same 100% uptime guarantees (for the services that support it).

Reply:
Sounds good. Does any of your commentary here belong in the doc itself?

Reply:
I've added most of the last paragraph, just for record-keeping